### ISSUE TITLE
IDE Integration

### ISSUE THREAD

**Issue by wborn** [openhab/static-code-analysis#333]
As a contributor it would be easier to comply to the SAT rules when they can be checked while using the openHAB Eclipse IDE. That way you don't have to run Maven commands and check SAT reports. You'd just see the PMD/Spotbugs/Checkstyle results as markers inline with your code.

All these plugins have IDE integrations. So we could:

1. Add some documentation on how to setup these tools such that they generate the SAT results.
2. Provide the plugins/configurations by default in the openHAB IDE.
3. Create a single IDE plugin which shows the results of all tools used by SAT combined.

**Comment by kaikreuzer** [comment]
4. Provide the plugins/configurations by default in the ESH IDE.

**Comment by lpapazow** [comment]
I researched how we could create such plugin and I saw that the eclipse checkstyle plugin does not provide the option to be called from another plugin. The checkstyle plugin only provides extension points for adding [custom checks, custom rulesets and custom filters](https://checkstyle.org/eclipse-cs/#!/extensions). The [pmd-eclipe-plugin provides extension point for custom ruleset](https://github.com/pmd/pmd-eclipse-plugin/blob/master/net.sourceforge.pmd.eclipse.plugin/plugin.xml)

Since all of our custom checks except for one are written using the checkstyle API, it would be possible to generate the same results in maven and eclipse builds. However, we would have to manually run checkstyle, PMD and then the SpotBugs plugins one by one and thus

> 3. Create a single IDE plugin which shows the results of all tools used by SAT combined.

will not be possible.

Alternatively, I saw that SonarQube allows [writing custom Java rules](https://docs.sonarqube.org/display/PLUG/Writing+Custom+Java+Rules+101) and allows integrating [checstyle](https://docs.sonarqube.org/display/SONARQUBE45/Checkstyle+Plugin), [pmd](https://docs.sonarqube.org/display/SONARQUBE45/PMD+Plugin) and [findbugs](https://docs.sonarqube.org/display/SONARQUBE45/FindBugs+Plugin) plugins as well. The [SonarLint](https://www.sonarlint.org/eclipse/) eclipse plugin could be connected to a SonarQube server, and will allow all of our maven checks (and more) to be executed by just one plugin. This option would require a running SonarQube server and would be very convinient if the ESH project is planning on introducing SonarQube as part of the build.

These are the possibilities I can think of and IMHO 1 and 2 are the best options:

1. When building the openhab.sat-plugin to create maven-plugin jar as well as eclipse-plugin jar that use the same custom rules/custom configurations. This would not satisfy the requirement to
> Create a single IDE plugin which shows the results of all tools used by SAT combined.

2. Introduce SonarQube for the ESH project and use SonarLint plugin.

3. Use the data from the maven generated reports to show errors in eclipse. The problem is that It would require a new maven run for changes to take effect.

@kaikreuzer  @martinvw  @wborn  do you like any of these suggestions or can you think of something else?

**Comment by wborn** [comment]
I don't think we should spend time on writing our own plugins. I've created https://github.com/openhab/static-code-analysis/issues/578 after doing a bit of research into what is possible with SARIF output nowadays.
